metricflow/test/model/validations/test_materializations.py

```diff
@@ -0,0 +1,280 @@
+    model = copy.deepcopy(simple_model__pre_transforms)
+    func_for_materialization: Callable[[Materialization], bool] = (
+        lambda materialization: len(materialization.destinations) > 0
+    )
+    materialization_with_destinations, _ = find_materialization_with(model, func_for_materialization)
+    assert type(materialization_with_destinations) == Materialization
+
+    # We assume the primary time dimension is `ds`
+    primary_time = "ds"
+
+    # We want to ensure the primary time dimension isn't used anywhere
+    destination: MaterializationDestination
+    for destination in materialization_with_destinations.destinations:
+        if destination.rollups:
+            new_rollups = []
+            for rollup_list in destination.rollups:
+                new_rollup_list = [
+                    name
+                    for name in rollup_list
+                    if StructuredLinkableSpecName.from_name(name).qualified_name_without_granularity != primary_time
+                ]
+                new_rollups.append(new_rollup_list)
+            destination.rollups = new_rollups
+
+    with pytest.raises(
+        ModelValidationException,
+        match=rf"Primary time dimension `{primary_time}` not listed as a ",
+    ):
+        ModelValidator.checked_validations(model)
+
+
+def test_primary_time_dimensions_are_granulated(simple_model__pre_transforms: UserConfiguredModel) -> None:  # noqa: D
+    model = copy.deepcopy(simple_model__pre_transforms)
+    func_for_materialization: Callable[[Materialization], bool] = (
+        lambda materialization: len(materialization.destinations) > 0
+    )
+    mat, _ = find_materialization_with(model, func_for_materialization)
+    assert mat is not None
+
+    # We assume the primary time dimension is `ds`
+    primary_time = StructuredLinkableSpecName(
+        identifier_link_names=(), element_name="ds", time_granularity=TimeGranularity.MONTH
+    )
+
+    # We want to ensure the primary time dimension isn't used anywhere
+    destination: MaterializationDestination
+    for destination in mat.destinations:
+        if destination.rollups:
+            new_rollups = []
+            for rollup_list in destination.rollups:
+                new_rollup_list = [
+                    name
+                    for name in rollup_list
+                    if StructuredLinkableSpecName.from_name(name).qualified_name_without_granularity
+                    != primary_time.qualified_name_without_granularity
+                ]
+                new_rollup_list.append(primary_time.qualified_name)
+                new_rollups.append(new_rollup_list)
+            destination.rollups = new_rollups
+
+    mat.dimensions = [
+        dim
+        for dim in mat.dimensions
+        if StructuredLinkableSpecName.from_name(dim).qualified_name_without_granularity
+        != primary_time.qualified_name_without_granularity
+    ]
+    mat.dimensions.append(primary_time.qualified_name)
```

Comment:
Why do we need this validation in the first place? Is it because materialization dimension references are user-provided dundered strings? If so, ick, but I guess it's what we're doing right now so we have to roll with it. I would prefer to just not give users the opportunity to forget to append granularity. Can we somehow have a standard validation pass that does this?

Also, we probably don't want to special-case primary time there. We need every time dimension to have granularity specified, right?